pkg/validation/spec/swagger_test.go

```diff
@@ -126,3 +135,126 @@ func TestSwaggerSpec_Deserialize(t *testing.T) {
 }
+
+func TestSwaggerRoundtrip(t *testing.T) {
+	cases := []jsontesting.RoundTripTestCase{
+		{
+			// Show at least one field from each embededd struct sitll allows
+			// roundtrips successfully
+			Name: "UnmarshalEmbedded",
+			Object: &Swagger{
+				VendorExtensible{Extensions{
+					"x-framework": "go-swagger",
+				}},
+				SwaggerProps{
+					Swagger: "2.0.0",
+				},
+			},
+		}, {
+			Name:   "BasicCase",
+			JSON:   specJSON,
+			Object: &spec,
+		},
+	}
+
+	for _, tcase := range cases {
+		t.Run(tcase.Name, func(t *testing.T) {
+			require.NoError(t, tcase.RoundTripTest(&Swagger{}))
+		})
+	}
+}
+
+func TestSwaggerSpec_ExperimentalUnmarshal(t *testing.T) {
+	fuzzer := fuzz.
+		NewWithSeed(1646791953).
+		NilChance(0.01).
+		MaxDepth(10).
+		NumElements(1, 2)
+
+	fuzzer.Funcs(
+		SwaggerFuzzFuncs...,
+	)
+
+	expected := Swagger{}
+	fuzzer.Fuzz(&expected)
+
+	// Serialize into JSON
+	jsonBytes, err := json.Marshal(expected)
+	require.NoError(t, err)
+
+	t.Log("Specimen", string(jsonBytes))
+
+	actual := Swagger{}
+	internal.UseExperimentalJSONUnmarshaling = true
+	err = json.Unmarshal(jsonBytes, &actual)
+	require.NoError(t, err)
+
+	if !reflect.DeepEqual(expected, actual) {
+		t.Fatal(cmp.Diff(expected, actual, SwaggerDiffOptions...))
+	}
+
+	control := Swagger{}
+	internal.UseExperimentalJSONUnmarshaling = false
+	err = json.Unmarshal(jsonBytes, &control)
+	require.NoError(t, err)
+
+	if !reflect.DeepEqual(control, actual) {
+		t.Fatal(cmp.Diff(control, actual, SwaggerDiffOptions...))
+	}
+
+	newJsonBytes, err := json.Marshal(actual)
+	require.NoError(t, err)
+	if !reflect.DeepEqual(jsonBytes, newJsonBytes) {
+		t.Fatal(cmp.Diff(string(jsonBytes), string(newJsonBytes), SwaggerDiffOptions...))
+	}
+}
+
+func BenchmarkSwaggerSpec_ExperimentalUnmarshal(b *testing.B) {
+	// Download kube-openapi swagger json
+	swagFile, err := os.Open("../../schemaconv/testdata/swagger.json")
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer swagFile.Close()
+
+	originalJSON, err := io.ReadAll(swagFile)
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	b.ResetTimer()
+
+	// Parse into kube-openapi types
+	b.Run("jsonv1", func(b2 *testing.B) {
+		internal.UseExperimentalJSONUnmarshaling = false
+		for i := 0; i < b2.N; i++ {
+			var result *Swagger
+			if err := json.Unmarshal(originalJSON, &result); err != nil {
+				b2.Fatal(err)
+			}
+		}
+	})
+
+	b.Run("jsonv2 via jsonv1", func(b2 *testing.B) {
+		internal.UseExperimentalJSONUnmarshaling = true
+		for i := 0; i < b2.N; i++ {
+			var result *Swagger
+			if err := json.Unmarshal(originalJSON, &result); err != nil {
+				b2.Fatal(err)
+			}
+		}
+	})
+
+	// Our UnmarshalJSON implementation which defers to jsonv2 causes the
+	// text to be parsed/validated twice. This costs a significant amount of time.
+	b.Run("jsonv2", func(b2 *testing.B) {
+		internal.UseExperimentalJSONUnmarshaling = true
+		for i := 0; i < b2.N; i++ {
+			var result *Swagger
+			if err := jsonv2.Unmarshal(originalJSON, &result); err != nil {
```

Comment:
Minor nit, slightly irrelevant: can you actually decode the way k8s will decode this, i.e.
```suggestion
			var result Swagger
			if err := result.UnmarshalJSON(originalJSON); err != nil {
```